website/blog/2023-10-21-incremental-decoding/index.md

```diff
@@ -0,0 +1,71 @@
+
+
+## Common Decoding Methods
+
+Here's an example to facilitate understanding different decoding methods:
+
+Let's say a developer starts to write a list comprehension in Python to get even numbers from a list:
+
+```python
+numbers = [1, 2, 3, 4, 5]
+evens = [x for x in numbers
+```
+
+To simplify the scenario, we assume that the language model maintains a probaility distribution as shown below,
+
+![probability](./probability.png)
+
+Here's how different decoding methods might suggest 🔍:
+
+
+### Greedy Decoding 🏆
+
+Greedy decoding selects the most probable next token at each step, which is most intuitive method but can sometimes lead to sub-optimal sequences. This is because it only considers one token at a time and makes such choice greedily. 
+
+In this particular case, greedy decoding would complete the code with `]\n print` as each of the token here has maximum probability given the chosen token before.
+
+![greedy](./greedy.png)
+
+### Beam Search 🌈
+Beam search maintains multiple possible sequences (beams) and expands them, which offers better quality at the expense of higher computation.
+
+Assuming `num_beams=2`, in this case, it'll actually produce ` if x % 2`, as `0.4*0.7=0.28` gives the highest probability when considering a sequence of 2.
+
+![beam](./beam.png)
+
+### Sampling-based methods 🎲
+The two methods above alway produce deterministic results given the language model probability distribution. Often times this isn't the ideal case, especially in conversational use case where users often retry to expect an alternative answer (or think about language translation). Alternatively, sampling-based methods like random sampling, top-k, and top-p sampling introduce randomness to achieve diverse outputs. 
+
+However, as it's now an undeterministic approach, sometimes the models could generate incoherent gibberish results. There are [many different sampling methods](https://huggingface.co/docs/transformers/v4.34.1/en/main_classes/text_generation#transformers.GenerationMixin.sample) to sharp the distribution or redistribute the probability mass to ensure higher chance of generating meaningful tasks. Here we won't go into the details.
+
+## Era of Streaming for LLM
+Latency is key in user experience for many LLM applications. In order to minmize the idle time for users, **streaming response** is commonly adopted. In LLM streaming, we start decoding the response as soon as it's available, instead of waiting for the entier response to be returned. 
+
+Considering streaming process in LLM decoding, although greedy decoding often produces sub-optimal results compared to beam decoding or sampling-methods methods, it wins with its fast and parallelizable computation.
```

Comment:
Probably emphasis that most LLM application today uses greedy search with certain sampling (e.g chatgpt, bard, antropic, ...)